### PR TITLE
Gemm-based packing ks

### PR DESCRIPTION
**Pull request by andrei-stoian-zama**
- Adds an optimized GEMM kernel
- Adds an alternative packing keyswitch implementation using the gemm kernel
- Implements a fast path for packing keyswitch for levels==1 using the gemm kernel
- Adds the fast path to the integer compression GPU code and adds a test with custom parameters that use the fast path

Achieves ~4x speedup for packing keyswitch

### CODE REVIEW

File: backends/tfhe-cuda-backend/cuda/src/crypto/fast_packing_keyswitch.cuh [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+__host__ void host_fast_packing_keyswitch_lwe_list_to_glwe(
+    cudaStream_t stream, uint32_t gpu_index, Torus *glwe_out,
+    Torus const *lwe_array_in, Torus const *fp_ksk_array, int8_t *fp_ks_buffer,
+    uint32_t lwe_dimension_in, uint32_t glwe_dimension,
+    uint32_t polynomial_size, uint32_t base_log, uint32_t level_count,
+    uint32_t num_lwes) {
+
+  // Optimization of packing keyswitch when packing many LWEs
+
+  printf("USING FAST PKS\n");
```
Comment (andrei-stoian-zama) [inline backends/tfhe-cuda-backend/cuda/src/crypto/fast_packing_keyswitch.cuh:177]: to be removed

File: tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs [hunk 2]
```diff
@@ -376,6 +376,7 @@ pub fn keyswitch_lwe_ciphertext_list_and_pack_in_glwe_ciphertext<
             .for_each(|mut poly| {
                 polynomial_wrapping_monic_monomial_mul_assign(&mut poly, MonomialDegree(degree));
             });
+
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/algorithms/lwe_packing_keyswitch.rs:379]: 🧐

File: tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs [hunk 3]
```diff
@@ -717,4 +727,145 @@ mod tests {
             }
         }
     }
+
+    #[test]
+    fn test_gpu_ciphertext_compression_fast_path() {
+        // these parameters are insecure
+        const PARAM_CUSTOM_FAST_PATH: ClassicPBSParameters = ClassicPBSParameters {
```
Comment (andrei-stoian-zama) [inline tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs:734]: @IceTDrinker , @agnesLeroy here I added dummy parameters that use the fast path but have no correctness guarantee - thus I removed the `and` operation in the test so no PBS is done -> just encrypt/compress/extract/decrypt. 

I hope this is ok, let me know.

File: tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs [hunk 4]
```diff
@@ -717,4 +727,145 @@ mod tests {
...
+            lwe_dimension: LweDimension(2048),
+            glwe_dimension: GlweDimension(1),
+            polynomial_size: PolynomialSize(2048),
```
Comment (andrei-stoian-zama) [inline tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs:737]: should I add other poly sizes in the test ?

**Comment by IceTDrinker** [comment]
hey I've been very busy with noise verifications, I should have some time to review hopefully in the next few days, but I have a ZK review that will take priority

**Comment by andrei-stoian-zama** [comment]
Closed because I generalized this approach in https://github.com/zama-ai/tfhe-rs/pull/1875

**Pull request closed without merging**